### PR TITLE
recording database updates

### PR DESCRIPTION
**Pull request by tsahee**
This converges support needed for validation without an archive node.

### CODE REVIEW

File: arbitrum/recordingdb.go [hunk 1]
```diff
@@ -167,29 +244,94 @@ func PrepareRecording(blockchain *core.BlockChain, lastBlockHeader *types.Header
...
+		currentHeader = r.bc.GetHeader(currentHeader.ParentHash, currentHeader.Number.Uint64()-1)
+		if currentHeader == nil {
+			return nil, fmt.Errorf("chain doesn't contain parent of block %d hash %v", currentHeader.Number, currentHeader.Hash())
+		}
+		stateDb, err = r.StateFor(currentHeader)
+		if err == nil {
+			lastRoot = currentHeader.Root
+			defer func() {
+				if (lastRoot != common.Hash{}) {
+					r.dereferenceRoot(lastRoot)
```
Comment (PlasmaPower) [inline arbitrum/recordingdb.go:297]: Let's move this defer out of the for loop to make it a bit more visible and clarify when it'll run

File: arbitrum/recordingdb.go [hunk 2]
```diff
@@ -167,29 +244,94 @@ func PrepareRecording(blockchain *core.BlockChain, lastBlockHeader *types.Header
...
+				}
+			}()
+			break
+		}
+	}
+	blockToRecreate := currentHeader.Number.Uint64() + 1
+	for ctx.Err() == nil {
+		block := r.bc.GetBlockByNumber(blockToRecreate)
+		if block == nil {
+			return nil, fmt.Errorf("block not found while recreating: %d", blockToRecreate)
```
Comment (PlasmaPower) [inline arbitrum/recordingdb.go:307]: We could check here if `block.ParentHash()` matches what we'd expect to provide early protection against reorgs, though I suppose checking that the state root matches is also good enough

**Review by PlasmaPower — COMMENTED** [review]
I have a couple minor comments but I love the approach!

File: arbitrum/recordingdb.go (hunk 2 above)
Comment (tsahee) [reply arbitrum/recordingdb.go:308]: done. Worth noting, though, that most reorgs will only be found after we reach the block number we're originally looking for and comparing it with the param header

**Review by PlasmaPower — APPROVED** [review]
LGTM

**Pull request merged**